### PR TITLE
[SBC-870]refactor to using custom RuleExt, not using Robin IRule.

### PR DESCRIPTION
**Pull request by liubao68**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by liubao68** [comment]
1. refactor to RuleExt
2. fix a bug required=true that int will have default value
3. This commit is a initial commit, I am going to add more tests

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/WeightedResponseTimeRuleExt.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  private static final int MIN_GAP = 10;
+
+  private Random random = new Random();
+
+  @Override
+  public ServiceCombServer choose(List<ServiceCombServer> servers, Invocation invocation) {
+    List<AtomicInteger> stats = new ArrayList<>(servers.size());
+    int totalWeights = 0;
+    boolean needRandom = false;
+    for (ServiceCombServer server : servers) {
```
Comment (jeho0815) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/WeightedResponseTimeRuleExt.java]: 每次都进行重新计算，如果server很多的时候，会影响性能吧

Comment (liubao68) [reply handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/WeightedResponseTimeRuleExt.java]: 权重算法就是这样的。对我们的场景，这个计算量很少，以及过滤掉不可用实例了。 即使200实例，这个算法复杂度也很低。

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadbalanceHandler.java [hunk 2]
```diff
@@ -148,16 +146,13 @@ public LoadbalanceHandler() {
...
-    if (!isRuleNotChanged) {
+    if (!isEqual(strategy, this.strategy)) {
       //配置变化，需要重新生成所有的lb实例
       synchronized (lock) {
         clearLoadBalancer();
       }
     }
-    this.policy = policy;
     this.strategy = strategy;
     LoadBalancer loadBalancer = getOrCreateLoadBalancer(invocation);
```
Comment (wujimin) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadbalanceHandler.java]: better to make loadBalancer statless
1.invoke discoveryTree before getOrCreateLoadBalancer
2.getOrCreateLoadBalancer
3.send or sendWithRetry with server list
  chooseServer with server list
  create RetryLoadBalancer instance with server list

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadbalanceHandler.java [hunk 3]
```diff
@@ -148,16 +146,13 @@ public LoadbalanceHandler() {
 
   @Override
   public void handle(Invocation invocation, AsyncResponse asyncResp) throws Exception {
-    String policy = Configuration.INSTANCE.getPolicy(invocation.getMicroserviceName());
     String strategy = Configuration.INSTANCE.getRuleStrategyName(invocation.getMicroserviceName());
-    boolean isRuleNotChanged = isEqual(policy, this.policy) && isEqual(strategy, this.strategy);
-    if (!isRuleNotChanged) {
+    if (!isEqual(strategy, this.strategy)) {
```
Comment (wujimin) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadbalanceHandler.java]: it's better to check if customers being using old policy configuration, and tell them how to switch to new mechanism
check can be done when init, not when handle
currently maybe init method not invoked.

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/RuleNameExtentionsFactory.java [hunk 4]
```diff
@@ -51,13 +47,13 @@ public boolean isSupport(String key, String value) {
   }
 
   @Override
-  public IRule createLoadBalancerRule(String ruleName) {
+  public RuleExt createLoadBalancerRule(String ruleName) {
     if (RULE_RoundRobin.equals(ruleName)) {
```
Comment (wujimin) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/RuleNameExtentionsFactory.java]: support SPI?
and use a map to manage them?

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/WeightedResponseTimeRuleExt.java (hunk 1 above)
Comment (wujimin) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/WeightedResponseTimeRuleExt.java]: calc weight for every invocation is not a good idea

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/RuleNameExtentionsFactory.java (hunk 4 above)
Comment (liubao68) [reply handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/RuleNameExtentionsFactory.java]: Now the extension point is ExtensionsFactory, not RuleExt.

**Pull request merged**